pkg/porter/publish.go

```diff
@@ -61,13 +101,151 @@ func (p *Porter) Publish(opts PublishOptions) error {
+// and application) listed and rename each based on the registry/org values derived from the provided tag.
+//
+// For each new image name, we use a library to copy the old image to the new image repository.
+// (Currently we use the pivotal/image-relocation library for this logic.)
+//
+// Finally, we generate a new bundle from the old, with all image names and digests updated, based
+// on the newly copied images, and then push this new bundle using the provided tag.
+// (Currently we use the docker/cnab-to-oci library for this logic.)
+//
+// In the generation of a new bundle, we therefore don't preserve content digests and can't maintain
+// signature verification throughout the process.  Once we wish to preserve content digest and such verification,
+// this approach will need to be refactored, via preserving the original bundle and employing
+// a relocation mapping approach to associate the bundle's (old) images with the newly copied images.
+func (p *Porter) publishFromArchive(opts PublishOptions) error {
+	if p.Debug {
+		fmt.Fprintf(p.Err, "Extracting bundle from archive %s...\n", opts.ArchiveFile)
+	}
+	source, err := filepath.Abs(opts.ArchiveFile)
+	if err != nil {
+		return errors.Wrapf(err, "could not determine absolute path to archive file %s", opts.ArchiveFile)
+	}
+
+	tmpDir, err := p.FileSystem.TempDir("", "porter")
+	if err != nil {
+		return errors.Wrap(err, "error creating temp directory for archive extraction")
+	}
+	defer p.FileSystem.RemoveAll(tmpDir)
+
+	l := loader.NewLoader()
+	imp := packager.NewImporter(source, tmpDir, l)
+
+	err = imp.Import()
+	if err != nil {
+		return errors.Wrapf(err, "failed to extract bundle from archive %s", opts.ArchiveFile)
+	}
+
+	bun, err := l.Load(filepath.Join(tmpDir, strings.TrimSuffix(filepath.Base(source), ".tgz"), "bundle.json"))
+	if err != nil {
+		return errors.Wrapf(err, "failed to load bundle from archive %s", opts.ArchiveFile)
+	}
+
+	if p.Debug {
+		fmt.Fprintf(p.Err, "Publishing bundle %s with tag %s...\n", bun.Name, opts.Tag)
+	}
+
+	// Update the bundle with new images (name, digest) based on the original,
+	// using the provided bundle tag to derive registry and org
+	for i, invImg := range bun.InvocationImages {
+		err := p.updateBundleWithNewImage(bun, invImg.Image, opts.Tag, i)
+		if err != nil {
+			return err
+		}
+	}
+	for name, img := range bun.Images {
+		err := p.updateBundleWithNewImage(bun, img.Image, opts.Tag, name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return p.Registry.PushBundle(bun, opts.Tag, opts.InsecureRegistry)
+}
+
+// updateBundleWithNewImage updates a bundle with a new image at the provided index
+// constructed using the provided original image and new bundle tag
+func (p *Porter) updateBundleWithNewImage(bun *bundle.Bundle, img, tag string, index interface{}) error {
+	newImg, err := getNewImageNameFromBundleTag(img, tag)
+	if err != nil {
+		return err
+	}
+
+	digest, err := p.Registry.Copy(img, newImg)
```

Comment:
So when we publish an archived bundle, the contents of the bundle don't really matter (the referenced images). We use the tags from where the images were originally published, and rely on being able to copy from there to the new location. We don't copy from the archive itself. Am I understanding that right?

Reply:
Good clarification -- indeed, this is the current approach.  

As I understand it, the functionality for importing images from an OCI Image layout (i.e., images in the `/artifacts` dir of the archive) into the Docker daemon is not yet available (see https://github.com/deislabs/cnab-go/issues/136)

Should I add a note/comment in-line mentioning this?  Would it be prudent to add a ticket in our backlog to revisit this area, If/when the aforementioned functionality is available?  (To import images straight from the archive into the local daemon and then re-tag/push.)

Reply:
So how does this work with the offline scenario where the bundle is on an air-gapped network?

I *think* we need to use the Pivotal relocation library here and copy *FROM* the archive up to a new registry. Loading things into the local docker daemon is nice, but for any case that involves a registry, it is kind of irrelevant isn't it?

Reply:
Mmm good point.  I'll work on the layout relocation approach directly from the archive and update.

Reply:
Ok @carolynvs & @jeremyrickard , updated logic to use the relocation lib to parse extracted OCI Layout and push renamed imgs directly from the archived images (same digests).  Did find one issue which I'll raise in the image-relocation repo, but I don't think it is a blocker (and I believe the fix would be in this dependency, anyhow).